### ISSUE TITLE
Getters throw exception if property is empty

### ISSUE THREAD

**Issue by Healyhatman** [calcinai/xero-php#652]
Trying to develop something where I need to update the PayTemplate of an Employee in PayrollAU. 

If the employee doesn't have a PayTemplate set already, calling getPayTemplate throws an Exception "Undefined index: PayTemplate". I assume this would happen with other properties as well.

**Possible Solutions:**
Set array properties to an empty array on initialisation of the object
Return the array property, or a new empty array
Add "hasX" methods that return a boolean, e.g. $employee->hasPayTemplate()

**Comment by calcinai** [comment]
Adding a ->hasX seems like a good solution.  The only current way to achieve this is with a `isset($employee->PayTemplate)`

**Comment by calcinai** [comment]
^^ Which is inconsistent, I know.

**Comment by Healyhatman** [comment]
Is that something I should do one by one and create a pull request for, or
is there an automagical thing you can do to apply it to all the classes?

[image: photo]
*Kamron Brooks*
Owner, Sky Spider
0438 526 766 <0438526766>
 skyspider.com.au
<https://s.codepen.io/boomerang/iFrameKey-193c05ea-7d6a-597e-e47f-90db67ddedfd/skyspider.com.au>

7 North Street, Teralba, NSW
<https://www.facebook.com/SkySpiderAus>
<https://www.linkedin.com/in/kamron-brooks-08160872/>

**Comment by Healyhatman** [comment]
I would much prefer if possible all the properties are set to empty
versions when the model is initialised though, so then you don't HAVE to do
->hasX before everything. Just having ->getPayTemplate() return an empty
PayTemplate makes more  sense to me than having to do
if($employee->hasPayTemplate()) { $tmpl = $employee->getPayTemplate()) else
{ $tmpl = new PayTemplate($xero)); }


[image: photo]
*Kamron Brooks*
Owner, Sky Spider
0438 526 766 <0438526766>
 skyspider.com.au
<https://s.codepen.io/boomerang/iFrameKey-193c05ea-7d6a-597e-e47f-90db67ddedfd/skyspider.com.au>

7 North Street, Teralba, NSW
<https://www.facebook.com/SkySpiderAus>
<https://www.linkedin.com/in/kamron-brooks-08160872/>

**Comment by calcinai** [comment]
Yeah, that seems like a reasonable approach.  I'm not sure why this wasn't default behaviour, but would be easily implemented in the base model.  All that we'd need to be aware of would be the impact of deliberately not posting properties that haven't been explicitly set.

**Comment by Healyhatman** [comment]
Can I close this? Can't remember what you ended up doing with it

**Comment by glennschmidt** [comment]
I can confirm that it's still an issue. Calling a getter for an uninitialized scalar property will generate an 'Undefined index' error/notice. Some _collection_ getters will initialize on demand, such as this one:

```
    /**
     * @return LineItem[]|Remote\Collection
     */
    public function getLineItems()
    {
        if (! isset($this->_data['LineItems'])) {
            $this->_data['LineItems'] = new Remote\Collection();
        }

        return $this->_data['LineItems'];
    }
```

but others don't:

```
    /**
     * @return Address[]|Remote\Collection
     */
    public function getAddresses()
    {
        return $this->_data['Addresses'];
    }
```

**Comment by calcinai** [comment]
@glennschmidt Interesting. I'm going o put that down to the fact all the models used to be generated, so every block was the same.  With the manual updates this must have slipped by.  Feel free to make a PR if you've got time to fix this?
